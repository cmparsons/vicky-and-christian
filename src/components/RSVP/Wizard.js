import React from "react";
import { Formik } from "formik";
import { Form, Button } from "semantic-ui-react";

export default class Wizard extends React.Component {
  static Page = ({ children }) => children;

  constructor(props) {
    super(props);
    this.state = {
      page: 0,
      values: props.initialValues
    };
  }

  next = values => {
    this.setState(state => ({
      page: Math.min(state.page + 1, this.props.children.length - 1),
      values
    }));
  };

  previous = () => {
    this.setState(state => ({
      page: Math.max(state.page - 1, 0)
    }));
  };

  validate = values => {
    const activePage = React.Children.toArray(this.props.children)[
      this.state.page
    ];
    return activePage.props.validate ? activePage.props.validate(values) : {};
  };

  handleSubmit = (values, bag) => {
    const { children, onSubmit } = this.props;
    const { page } = this.state;
    const isLastPage = page === React.Children.count(children) - 1;
    if (isLastPage) {
      return onSubmit(values, bag);
    } else {
      bag.setTouched({});
      bag.setSubmitting(false);
      this.next(values);
    }
  };

  render() {
    const { children } = this.props;
    const { page, values } = this.state;
    const activePage = React.Children.toArray(children)[page];
    const isLastPage = page === React.Children.count(children) - 1;
    return (
      <Formik
        initialValues={values}
        enableReinitialize={false}
        validate={this.validate}
        onSubmit={this.handleSubmit}
        render={({ handleSubmit, isSubmitting }) => (
          <Form size="large" onSubmit={handleSubmit} loading={isSubmitting}>
            {activePage}
            <>
              {page > 0 && (
                <Button secondary type="button" onClick={this.previous}>
                  Prev
                </Button>
              )}
              {isLastPage ? (
                <Button color="green" type="submit">
                  Submit
                </Button>
              ) : (
                <Button primary type="submit">
                  Next
                </Button>
              )}
            </>
          </Form>
        )}
      />
    );
  }
}
