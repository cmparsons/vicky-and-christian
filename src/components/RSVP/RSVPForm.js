import React from "react";
import { Form, Popup, Radio, TextArea } from "semantic-ui-react";
import { Field } from "formik";
import Wizard from "./Wizard";

const wait = (delay = 2000) =>
  new Promise(resolve => setTimeout(resolve, delay));

export default function RSVPForm() {
  return (
    <Wizard
      initialValues={{
        firstName: "",
        lastName: "",
        eventCode: "",
        isAttending: "",
        mailingAddress: "",
        contactInfo: ""
      }}
      onSubmit={(values, { setSubmitting }) => {
        values.isAttending = values.isAttending === "true";

        // const eventsRef = firebase.firestore().collection("event");
        // const querySnapshot = await eventsRef.get();

        // const [realEventCode] = querySnapshot.docs;

        // if (realEventCode.data().code === eventCode) {
        //   console.log("eventCode match");
        // } else {
        //   console.log("eventCode did not match");
        // }

        // const guestsRef = firebase.firestore().collection('guests');

        // const guests = guestsRef
        //   .where('firstName', '==', firstName)
        //   .where('lastName', '==', lastName);

        // const guestsQuerySnapshot = await guests.get();
        // guestsQuerySnapshot.docs

        wait().then(() => {
          console.log(values);
          setSubmitting(false);
        });
      }}
    >
      <Wizard.Page
        validate={values => {
          if (
            process.env.REACT_APP_EVENT_CODE &&
            values.eventCode !== process.env.REACT_APP_EVENT_CODE
          ) {
            return {
              eventCode: "Invalid event code"
            };
          }

          return {};
        }}
      >
        <Form.Group widths="equal">
          <Field
            name="firstName"
            render={({ field }) => (
              <Form.Input
                {...field}
                required
                placeholder="First Name"
                label="First name"
              />
            )}
          />
          <Field
            name="lastName"
            render={({ field }) => (
              <Form.Input
                {...field}
                required
                placeholder="Last Name"
                label="Last name"
              />
            )}
          />
        </Form.Group>
        <Field
          name="eventCode"
          render={({ field }) => (
            <Popup
              trigger={
                <Form.Input
                  {...field}
                  required
                  icon="lock"
                  iconPosition="left"
                  label="Event code"
                  placeholder="Event Code"
                />
              }
              label="Event Code"
              content="Check your invite to find the event code"
              on="focus"
            />
          )}
        />
      </Wizard.Page>
      <Wizard.Page>
        <label
          style={{
            display: "block",
            marginBottom: "15px",
            fontWeight: "bold"
          }}
        >
          Will you be attending the wedding?
        </label>
        <Field
          name="isAttending"
          render={({ field, form }) => (
            <Form.Field
              {...field}
              control={Radio}
              id="isAttending_true"
              name="isAttending"
              label="Yes"
              value="true"
              checked={field.value === "true"}
              error={form.touched.isAttending && form.errors.isAttending}
            />
          )}
        />
        <Field
          name="isAttending"
          render={({ field, form }) => (
            <Form.Field
              {...field}
              control={Radio}
              name="isAttending"
              id="isAttending_false"
              label="No"
              value="false"
              checked={field.value === "false"}
              error={form.touched.isAttending && form.errors.isAttending}
            />
          )}
        />
      </Wizard.Page>
      <Wizard.Page>
        <Field
          name="mailingAddress"
          render={({ field }) => (
            <Form.Field
              {...field}
              control={TextArea}
              label="What is your mailing address?"
              placeholder="Please enter your mailing address"
              rows={5}
            />
          )}
        />
      </Wizard.Page>
      <Wizard.Page>
        <Field
          name="contactInfo"
          render={({ field }) => (
            <Form.Field
              {...field}
              control={TextArea}
              label="What is a good way to contact you? (preferably a phone number or email)"
              placeholder="Please enter your contact information"
              rows={5}
            />
          )}
        />
      </Wizard.Page>
    </Wizard>
  );
}
