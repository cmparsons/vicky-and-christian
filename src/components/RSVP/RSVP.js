import React from "react";
import RSVPForm from "./RSVPForm";
import { Container, Grid, Header, Segment } from "semantic-ui-react";

export default function RSVP() {
  return (
    <Container>
      <Grid verticalAlign="middle" centered style={{ height: "100vh" }}>
        <Grid.Column style={{ maxWidth: 650 }}>
          <Header as="h2" textAlign="center">
            <span role="img" aria-label="RSVP">
              üìù
            </span>
            RSVP
          </Header>
          <Segment stacked>
            <RSVPForm />
          </Segment>
        </Grid.Column>
      </Grid>
    </Container>
  );
}
