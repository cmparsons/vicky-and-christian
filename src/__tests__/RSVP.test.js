import React from "react";
import { render, fireEvent, waitForDomChange } from "@testing-library/react";
import RSVP from "../components/RSVP/RSVP";

const generateString = length => "a".repeat(length);

test("wizard goes to next page when no errors on current page", async () => {
  const { getByLabelText, getByText, queryByLabelText, queryByText } = render(
    <RSVP />
  );

  // First page -- first name, last name
  fireEvent.change(getByLabelText(/first name/i), {
    target: { value: "test" }
  });
  fireEvent.change(getByLabelText(/last name/i), { target: { value: "user" } });
  fireEvent.click(getByText(/next/i));

  await waitForDomChange();

  expect(queryByLabelText(/first name/i)).not.toBeInTheDocument();
  expect(queryByLabelText(/last name/i)).not.toBeInTheDocument();

  // Second page -- is attending
  expect(queryByText(/attending/i)).toBeInTheDocument();
});

test("form values persist from previous page", async () => {
  const { getByLabelText, getByText, queryByLabelText } = render(<RSVP />);

  // First page -- first name, last name
  fireEvent.change(getByLabelText(/first name/i), {
    target: { value: "test" }
  });
  fireEvent.change(getByLabelText(/last name/i), { target: { value: "user" } });
  fireEvent.click(getByText(/next/i));

  await waitForDomChange();

  fireEvent.click(getByText(/prev/i));

  expect(queryByLabelText(/first name/i)).toHaveValue("test");
  expect(queryByLabelText(/last name/i)).toHaveValue("user");
});

test("requires user to fill out first name and last name", async () => {
  const { queryAllByText, getByText } = render(<RSVP />);

  // First page -- first name, last name
  fireEvent.click(getByText(/next/i));

  await waitForDomChange();

  expect(queryAllByText("Required")).toHaveLength(2);
});

test('requires user to fill out "Is attending?" question', async () => {
  const { getByLabelText, getByText, queryByText } = render(<RSVP />);

  // First page -- first name, last name
  fireEvent.change(getByLabelText(/first name/i), {
    target: { value: "first" }
  });
  fireEvent.change(getByLabelText(/last name/i), { target: { value: "last" } });
  fireEvent.click(getByText(/next/i));

  await waitForDomChange();

  // Second page -- is attending
  fireEvent.click(getByText(/next/i));

  await waitForDomChange();

  expect(queryByText("Required")).toBeInTheDocument();
});

test("prevents user from submitting long responses for mailing address", async () => {
  const { getByLabelText, getByText, queryByText } = render(<RSVP />);

  // First page -- first name, last name
  fireEvent.change(getByLabelText(/first name/i), {
    target: { value: "first" }
  });
  fireEvent.change(getByLabelText(/last name/i), { target: { value: "last" } });
  fireEvent.click(getByText(/next/i));
  await waitForDomChange();

  // Second page -- is attending
  fireEvent.click(getByLabelText(/yes/i));
  fireEvent.click(getByText(/next/i));
  await waitForDomChange();

  // Third page -- mailing address
  fireEvent.change(getByLabelText(/what is your mailing address?/i), {
    target: { value: generateString(1000) }
  });
  fireEvent.click(getByText(/next/i));
  await waitForDomChange();

  expect(queryByText(/response is too long/i)).toBeInTheDocument();
});

test("prevents user from submitting long responses for contact info", async () => {
  const { getByLabelText, getByText, queryByText } = render(<RSVP />);

  // First page -- first name, last name
  fireEvent.change(getByLabelText(/first name/i), {
    target: { value: "first" }
  });
  fireEvent.change(getByLabelText(/last name/i), { target: { value: "last" } });
  fireEvent.click(getByText(/next/i));
  await waitForDomChange();

  // Second page -- is attending
  fireEvent.click(getByLabelText(/yes/i));
  fireEvent.click(getByText(/next/i));
  await waitForDomChange();

  // Third page -- mailing address
  fireEvent.change(getByLabelText(/mailing address/i), {
    target: { value: "my mailing address" }
  });
  fireEvent.click(getByText(/next/i));
  await waitForDomChange();

  // Fourth page -- contact information
  fireEvent.change(getByLabelText(/contact/i), {
    target: { value: generateString(1000) }
  });
  fireEvent.click(getByText(/next/i));
  await waitForDomChange();

  expect(queryByText(/response is too long/i)).toBeInTheDocument();
});

test("allows the user to rsvp successfully", async () => {
  jest.spyOn(window, "fetch").mockImplementationOnce(async () => ({
    ok: true
  }));

  const { getByLabelText, getByText, findByRole } = render(<RSVP />);

  // First page -- first name, last name
  fireEvent.change(getByLabelText(/first name/i), {
    target: { value: "test" }
  });
  fireEvent.change(getByLabelText(/last name/i), { target: { value: "user" } });
  fireEvent.click(getByText(/next/i));
  await waitForDomChange();

  // Second page -- is attending
  fireEvent.click(getByLabelText(/yes/i));
  fireEvent.click(getByText(/next/i));
  await waitForDomChange();

  // Third page -- mailing address
  fireEvent.change(getByLabelText(/mailing address/i), {
    target: { value: "my mailing address" }
  });
  fireEvent.click(getByText(/next/i));
  await waitForDomChange();

  // Fourth page -- contact information
  fireEvent.change(getByLabelText(/contact/i), {
    target: { value: "my contact info" }
  });
  fireEvent.click(getByText(/next/i));
  await waitForDomChange();

  // Fifth page -- party members
  fireEvent.click(getByText(/submit/i));

  const message = await findByRole("alert");

  expect(message).toHaveTextContent(/success/i);
});

test("displays error if an API error occurs", async () => {
  jest.spyOn(window, "fetch").mockImplementationOnce(async () => ({
    status: 400
  }));

  const { getByLabelText, getByText, findByRole } = render(<RSVP />);

  // First page -- first name, last name
  fireEvent.change(getByLabelText(/first name/i), {
    target: { value: "test" }
  });
  fireEvent.change(getByLabelText(/last name/i), { target: { value: "user" } });
  fireEvent.click(getByText(/next/i));
  await waitForDomChange();

  // Second page -- is attending
  fireEvent.click(getByLabelText(/yes/i));
  fireEvent.click(getByText(/next/i));
  await waitForDomChange();

  // Third page -- mailing address
  fireEvent.change(getByLabelText(/mailing address/i), {
    target: { value: "my mailing address" }
  });
  fireEvent.click(getByText(/next/i));
  await waitForDomChange();

  // Fourth page -- contact information
  fireEvent.change(getByLabelText(/contact/i), {
    target: { value: "my contact info" }
  });
  fireEvent.click(getByText(/next/i));
  await waitForDomChange();

  // Fifth page -- party members
  fireEvent.click(getByText(/submit/i));

  const message = await findByRole("alert");

  expect(message).toHaveTextContent(/error/i);
});

test("first input of each page is focused when initially mounted", async () => {
  const { getByLabelText, getByText } = render(<RSVP />);

  // First page -- first name, last name
  const firstNameInput = getByLabelText(/first name/i);
  expect(firstNameInput).toHaveFocus();
  fireEvent.change(firstNameInput, { target: { value: "test" } });
  fireEvent.change(getByLabelText(/last name/i), { target: { value: "user" } });
  fireEvent.click(getByText(/next/i));

  await waitForDomChange();

  // Second page -- is attending
  // TODO ...
  fireEvent.click(getByLabelText(/yes/i));
  fireEvent.click(getByText(/next/i));
  await waitForDomChange();

  // Third page -- mailing address
  const mailingAddress = getByLabelText(/mailing address/i);
  expect(mailingAddress).toHaveFocus();
  fireEvent.click(getByText(/next/i));
  await waitForDomChange();

  // Fourth page -- contact info
  const contactInfo = getByLabelText(/contact/i);
  expect(contactInfo).toHaveFocus();
  fireEvent.click(getByText(/next/i));
  await waitForDomChange();

  // Fifth page -- party
  const firstPartyMemberFirstName = getByLabelText(/first name/i);
  expect(firstPartyMemberFirstName).toHaveFocus();
});

test("prevents user from adding multiple first names", async () => {
  const { getByLabelText, getByText } = render(<RSVP />);

  fireEvent.change(getByLabelText(/first name/i), {
    target: { value: "Bob and Mary" }
  });
  fireEvent.change(getByLabelText(/last name/i), {
    target: { value: "Smith" }
  });
  fireEvent.click(getByText(/next/i));

  await waitForDomChange();

  expect(getByText(/limit first name to one person/i)).toBeInTheDocument();
});
