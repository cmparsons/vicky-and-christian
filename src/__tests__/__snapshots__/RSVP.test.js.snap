// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`trims whitespace in form values before sending request 1`] = `
Array [
  "fake-firebase-functions.com/rsvp",
  Object {
    "body": "{\\"firstName\\":\\"test\\",\\"lastName\\":\\"user\\",\\"eventCode\\":\\"\\",\\"isAttending\\":true,\\"mailingAddress\\":\\"my mailing address\\",\\"contactInfo\\":\\"my contact info\\",\\"party\\":[]}",
    "headers": Object {
      "Content-Type": "application/json",
    },
    "method": "POST",
  },
]
`;
